server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java

```diff
@@ -313,13 +316,28 @@ protected void setMinSeqNoToKeep(long seqNo) {
         this.minSeqNoToKeep = seqNo;
     }
 
-    @Override
-    void deleteReaderFiles(TranslogReader reader) {
+    void deleteRemoteGeneration(long generation) {
         try {
-            translogTransferManager.deleteTranslog(primaryTermSupplier.getAsLong(), reader.generation);
+            translogTransferManager.deleteTranslog(primaryTermSupplier.getAsLong(), generation);
         } catch (IOException ignored) {
-            logger.error("Exception {} while deleting generation {}", ignored, reader.generation);
+            logger.error("Exception {} while deleting generation {}", ignored, generation);
+        }
+    }
+
+    public void trimUnreferencedReaders() throws IOException {
+        // clean up local translog files and updates readers
+        super.trimUnreferencedReaders();
+
+        // cleans up remote translog files not referenced in latest uploaded metadata
+        // This enables us to restore translog from that metadata
+        for (long generation = minRemoteGenReferenced - 1; generation > 0; generation--) {
+            String translogFilename = Translog.getFilename(generation);
+            if (fileTransferTracker.uploaded(translogFilename)) {
+                logger.trace("delete remote translog generation file [{}], not referenced by metadata anymore", generation);
+                deleteRemoteGeneration(generation);
```

Comment:
deleteRemoteGeneration method uses the current primaryTerm. This probably would not be handled when there is failover and the primary term has increased. Could we rely on metadata to fetch the right generation to primary term mapping for cleaning up the files?

Reply:
That's a great catch @ashking94 . The issue here is the latest metadata doesn't know about the right primary term for this generation. I will create a backlog item for cleaning up the older primary term translog files . This can be done on every failover or an async job . I don't want to complicate the usual deletion flow due to this corner case .